### PR TITLE
Add retry with exponential backoff on recoverable error

### LINKED ISSUE #71: We should retry the reconcile loop if we fail to create a secret
given secret dependencies, we may require more than one iteration of the reconcile loop to create all secrets. We should trigger a retry if we fail to create a secret. We should only retry a MAX number of times to avoid an infinite loop. Maybe 10?

### PR DESCRIPTION
**Pull request by jrcast**
1. Add retry with exponential backoff on recoverable error
1. Add new `State` to SAC Status
1. Update client-go module

Closes #71

### CODE REVIEW

File: controllers/secretagentconfiguration_controller.go [hunk 1]
```diff
@@ -229,6 +243,14 @@ func (reconciler *SecretAgentConfigurationReconciler) updateStatus(ctx context.C
 	instance.Status.ManagedK8sSecrets = secretNames
 	instance.Status.TotalManagedObjects = totalManagedObjects
 
+	if errorFound {
+		instance.Status.State = v1alpha1.SecretAgentConfigurationError
+	} else if inProgress {
+		instance.Status.State = v1alpha1.SecretAgentConfigurationInProgress
+	} else {
```
Comment (maxres-fr) [inline controllers/secretagentconfiguration_controller.go:250]: We default `errorFound` to false, if we miss catching some error or whatever in the reconcile loop this leads to a default state of "completed". I'd rather have an unknown or unhandled state as a safety, I think the "complete" state might lead to false positives.

File: go.mod [hunk 2]
```diff
@@ -12,16 +12,16 @@ require (
...
 	github.com/go-logr/logr v0.1.0
 	github.com/go-playground/validator/v10 v10.2.0
-	github.com/onsi/ginkgo v1.11.0
-	github.com/onsi/gomega v1.8.1
+	github.com/onsi/ginkgo v1.12.1
+	github.com/onsi/gomega v1.10.1
 	github.com/pkg/errors v0.9.1
 	go.uber.org/zap v1.10.0
 	golang.org/x/crypto v0.0.0-20200510223506-06a226fb4e37
+	golang.org/x/time v0.0.0-20191024005414-555d28b269f0
```
Comment (maxres-fr) [inline go.mod:20]: time isn't part of stdlib? _jusy say'n?_

File: api/v1alpha1/secretagentconfiguration_types.go [hunk 3]
```diff
@@ -70,6 +71,16 @@ type SecretAgentConfigurationList struct {
 	Items           []SecretAgentConfiguration `json:"items"`
 }
 
+// SecretAgentConfState is used to keep track of the SAC state
+type SecretAgentConfState string
+
+// SecretsManager Strings
+const (
```
Comment (maxres-fr) [inline api/v1alpha1/secretagentconfiguration_types.go:78]: See my comment in the updatestatus method re: statuses.

File: api/v1alpha1/secretagentconfiguration_types.go [hunk 4]
```diff
@@ -70,6 +71,16 @@ type SecretAgentConfigurationList struct {
...
+	SecretAgentConfigurationCompleted  SecretAgentConfState = "Completed"
+	SecretAgentConfigurationInProgress SecretAgentConfState = "InProgress"
+	SecretAgentConfigurationError      SecretAgentConfState = "ErrorFound"
+)
```
Comment (maxres-fr) [inline api/v1alpha1/secretagentconfiguration_types.go:82]:
```suggestion
const (
	SecretAgentConfigurationCompleted  SecretAgentConfState = "Completed"
	SecretAgentConfigurationInProgress SecretAgentConfState = "InProgress"
	SecretAgentConfigurationError      SecretAgentConfState = "ErrorFound"
	SecretAgentConfigurationMisconfigured      SecretAgentConfState = "Misconfigured"
)
```

File: controllers/secretagentconfiguration_controller.go (hunk 1 above)
Comment (maxres-fr) [reply controllers/secretagentconfiguration_controller.go:250]:
```suggestion
	} else if !errorFound && !inProggress {
			instance.Status.State = v1alpha1.SecretAgentConfigurationInProgress
	} else {
	instance.Status.State = v1alpah1.SecretAgentConfigurationMisconfigured
	}
```

Comment (jrcast) [reply controllers/secretagentconfiguration_controller.go:250]: we decided to keep it as is

**Pull request merged**